### PR TITLE
Export to Redshift from the Production Reporting RDS Proxy Endpoint

### PR DESCRIPTION
**Pull request by sureshc**
We have been cloning the production Aurora database cluster and then configuring our AWS Database Migration Service Replication Tasks to source data from the clone when carrying out our daily batch export to Redshift to reduce load on the production database. Switch to using the production reporting RDS Proxy Endpoint instead which should ensure no negative impact of the large read queries on the reader database instance performance.

We're doing this because #57815 changed the production database configuration in a way that breaks the rake tasks that clone clusters.  See [internal discussion](https://codedotorg.slack.com/archives/C03CK49G9/p1713984298941749).

Also - split the configuration settings for the Redshift admin SQL user and the Redshift SQL user utilized for ETL processes.

Note that the Stack Parameters `redshift_username` and `redshift_password` get populated automatically from AWS Secrets with matching names when executing the rake `stack` commands. Only the new admin Parameters need to be set using Environment variables when executing the `stack:data:start` task.  See execution of `stack:data:validate` below.

## Testing story

DONE - Manually checkout the modification to the cron job in this feature branch to the production-daemon server, manually configure the Database Migration Service Endpoint to point to the production reporting RDS Proxy endpoint, and manually execute the daily export job in a `screen` on `production-daemon` while monitoring production database performance.

Verify changes to the data stack (the interactive Parameter value prompting is not currently supported in the `validate` task, so provide placeholder values on the command line):

```
$ export AWS_PROFILE=codeorg-admin
$ bin/aws_access
AWS access: GoogleSignInAdmin/suresh@code.org
$ bundle exec rake stack:data:validate RAILS_ENV=production ADMIN=true REDSHIFT_ADMIN_USERNAME=[PLACEHOLDER] REDSHIFT_ADMIN_PASSWORD=[PLACEHOLDER] REDSHIFT_ETL_CLIENT_USERNAME=[PLACEHOLDER] REDSHIFT_ETL_CLIENT_PASSWORD=[PLACEHOLDER]

Pending update for stack `DATA-production`:
Pending update for stack `DATA-production`:
Remove AuroraCloneEndpoint [AWS::DMS::Endpoint]
Modify DMSCronLevelSourcesPii [AWS::DMS::ReplicationTask] Properties Replacement: Conditional (ReplicationTaskIdentifier, SourceEndpointArn, SourceEndpointArn, TargetEndpointArn)
Modify DMSCronPii [AWS::DMS::ReplicationTask] Properties Replacement: Conditional (ReplicationTaskIdentifier, SourceEndpointArn, SourceEndpointArn, TargetEndpointArn)
Modify DMSCronUserHierarchyPii [AWS::DMS::ReplicationTask] Properties Replacement: Conditional (ReplicationTaskIdentifier, SourceEndpointArn, SourceEndpointArn, TargetEndpointArn)
Modify DMSCronUserHierarchy [AWS::DMS::ReplicationTask] Properties Replacement: Conditional (ReplicationTaskIdentifier, SourceEndpointArn, SourceEndpointArn, TargetEndpointArn)
Modify DMSCronUserLevels [AWS::DMS::ReplicationTask] Properties Replacement: Conditional (ReplicationTaskIdentifier, SourceEndpointArn, SourceEndpointArn, TargetEndpointArn)
Modify DMSCron [AWS::DMS::ReplicationTask] Properties Replacement: Conditional (ReplicationTaskIdentifier, SourceEndpointArn, SourceEndpointArn, TargetEndpointArn)
Add MySQLDatabaseSourceEndpoint [AWS::DMS::Endpoint]
Remove RDSEndpoint [AWS::DMS::Endpoint]
Modify RedshiftEndpoint [AWS::DMS::Endpoint] Properties Replacement: Conditional (Password, Password, Port, Username, ServerName, Username)
Modify Tableau [AWS::Redshift::Cluster] Properties (MasterUserPassword, MasterUserPassword)
Finished stack:data:validate (less than a minute)
```

## Deployment strategy

An Infrastructure engineer must update the `DATA-production` CloudFormation Stack before these changes are released to production.

The `start` task interactively prompts for values of Parameters that don't currently have values which prevents specifying the values with environment variables and potentially saving sensitive information in command history:

```
$ export AWS_PROFILE=codeorg-admin
$ bin/aws_access
AWS access: GoogleSignInAdmin/suresh@code.org
$ bundle exec rake stack:data:start RAILS_ENV=production ADMIN=true
```

## Follow-up work

We also need to fix the database cluster cloning library to support cloning clusters that use the [default cluster ParameterGroup](https://github.com/code-dot-org/code-dot-org/blob/f280b95523fb6a3aabbe624250146256a9480cef/lib/cdo/aws/rds.rb#L30-L34) or [default database instance ParameterGroup](https://github.com/code-dot-org/code-dot-org/blob/f280b95523fb6a3aabbe624250146256a9480cef/lib/cdo/aws/rds.rb#L54-L58).

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Review by jordan-springer — APPROVED** [review]
LGTM

**Review by Hamms — APPROVED** [review]

File: aws/cloudformation/data.yml.erb [hunk 1]
```diff
@@ -97,27 +97,17 @@ Resources:
             Resource:
               - 'arn:aws:s3:::cdo-data-sharing-internal/*'
               - 'arn:aws:s3:::cdo-data-sharing-internal'
-  RDSEndpoint:
+  ProductionMySQLDatabaseReportingEndpoint:
```
Comment (cat5inthecradle) [inline aws/cloudformation/data.yml.erb]: Noticing the word `Production` here. What does this word mean in this context? Especially given the use of `AWS::StackName` inside of it.

File: aws/cloudformation/data.yml.erb [hunk 2]
```diff
@@ -97,27 +97,17 @@ Resources:
...
-    Type: AWS::DMS::Endpoint
-    Properties:
-      EndpointIdentifier: !Sub "AuroraCloneEndpoint-${AWS::StackName}"
+      EndpointIdentifier: !Sub "MySQLReportingEndpoint-${AWS::StackName}"
       EndpointType: source
       EngineName: aurora
-      # At the time this template is rendered the Redshift Import Clone may not exist, so we can't use the RDS describe
-      # endpoints API to get this server name.
-      ServerName: !Sub "<%= RedshiftImport::CLONE_CLUSTER_ID %>.${RDSClusterEndpointRootDomain}"
+      # The Stack Parameter RDSHost must be set to the domain name of the production RDS Proxy reporting endpoint.
```
Comment (cat5inthecradle) [inline aws/cloudformation/data.yml.erb]: Move/Copy this to the parameter description.

File: aws/cloudformation/data.yml.erb [hunk 3]
```diff
@@ -97,27 +97,17 @@ Resources:
...
+      ServerName: !Ref RDSHost
       Port: 3306
+      # The RDSUsername and RDSPassword Stack Parameters must be set to a read only user that can authenticate on the
```
Comment (cat5inthecradle) [inline aws/cloudformation/data.yml.erb:118]: Move/Copy this to the parameter description

File: cookbooks/cdo-apps/templates/default/crontab.erb [hunk 4]
```diff
@@ -119,7 +119,7 @@
       cronjob at:'5 12 * * *', do:deploy_dir('bin', 'cron', 'applab_datasets', 'daily_weather')
       cronjob at:'0 12 * * *', do:deploy_dir('bin', 'cron', 'applab_datasets', 'spotify')
       cronjob at:'0 12 * * *', do:deploy_dir('bin', 'cron', 'applab_datasets', 'covid19')
-      cronjob at:'00 02 * * *', do:deploy_dir('bin', 'cron', 'export_mysql_database_to_redshift')
+      cronjob at:'00 23 * * *', do:deploy_dir('bin', 'cron', 'export_mysql_database_to_redshift')
```
Comment (cat5inthecradle) [inline cookbooks/cdo-apps/templates/default/crontab.erb:122]: Why are we moving this earlier?

File: bin/cron/export_mysql_database_to_redshift [hunk 5]
```diff
@@ -19,13 +18,6 @@ WEEKLY_TASK_MAX_EXECUTION_TIME = 40.hours
 
 def main
   ChatClient.message 'data', 'Beginning daily export from Aurora MySQL database to Redshift.'
-  # Delete clone of production cluster if it was incorrectly left behind by a previous execution of the import process.
-  Cdo::RDS.delete_cluster(RedshiftImport::CLONE_CLUSTER_ID)
-  Cdo::RDS.clone_cluster(
-    source_cluster_id: CDO.db_cluster_id,
-    clone_cluster_id: RedshiftImport::CLONE_CLUSTER_ID,
-    instance_type: RedshiftImport::CLONE_DB_INSTANCE_ID
-  )
```
Comment (cat5inthecradle) [inline bin/cron/export_mysql_database_to_redshift:30]: Add a follow-up task to make sure the final clone was deleted.

**Review by cat5inthecradle — CHANGES_REQUESTED** [review]

File: cookbooks/cdo-apps/templates/default/crontab.erb (hunk 4 above)
Comment (sureshc) [reply cookbooks/cdo-apps/templates/default/crontab.erb:122]: To better align with our lowest usage time periods (~4PM PT to ~4AM PT)

File: aws/cloudformation/data.yml.erb (hunk 2 above)
Comment (sureshc) [reply aws/cloudformation/data.yml.erb]: Done!

File: aws/cloudformation/data.yml.erb (hunk 3 above)
Comment (sureshc) [reply aws/cloudformation/data.yml.erb:118]: Done!

File: aws/cloudformation/data.yml.erb (hunk 1 above)
Comment (sureshc) [reply aws/cloudformation/data.yml.erb]: renamed to `MySQLDatabaseSourceEndpoint`

**Review by cat5inthecradle — APPROVED** [review]

File: aws/cloudformation/data.yml.erb [hunk 6]
```diff
@@ -17,11 +17,15 @@ Parameters:
     Type: String
   # MySQL user must have granted REPLICATION CLIENT and REPLICATION SLAVE permissions.
   # Ref: http://docs.aws.amazon.com/dms/latest/userguide/CHAP_Source.MySQL.html#CHAP_Source.MySQL.Security
+  # and must have SELECT permissions on all tables and must have permissions to connect to the Production Reporting
+  # RDS Proxy.
```
Comment (Hamms) [inline aws/cloudformation/data.yml.erb]: nit:
```suggestion
  # MySQL user must have granted REPLICATION CLIENT and REPLICATION SLAVE
  # permissions, and must have SELECT permissions on all tables and must have
  # permissions to connect to the Production Reporting RDS Proxy.
  # Ref: http://docs.aws.amazon.com/dms/latest/userguide/CHAP_Source.MySQL.html#CHAP_Source.MySQL.Security
```

File: bin/cron/export_mysql_database_to_redshift [hunk 7]
```diff
@@ -21,13 +20,6 @@ NOTIFICATION_CHANNEL = 'data-alerts'
 
 def main
   ChatClient.message NOTIFICATION_CHANNEL, 'Beginning daily export from Aurora MySQL database to Redshift.'
-  # Delete clone of production cluster if it was incorrectly left behind by a previous execution of the import process.
-  Cdo::RDS.delete_cluster(RedshiftImport::CLONE_CLUSTER_ID)
-  Cdo::RDS.clone_cluster(
-    source_cluster_id: CDO.db_cluster_id,
-    clone_cluster_id: RedshiftImport::CLONE_CLUSTER_ID,
-    instance_type: RedshiftImport::CLONE_DB_INSTANCE_ID
```
Comment (Hamms) [inline bin/cron/export_mysql_database_to_redshift:29]: Can these constants also be removed from https://github.com/code-dot-org/code-dot-org/blob/94020ea309cfd32d189c686f5cd73065f117a6e5/lib/cdo/redshift_import.rb#L4-L5 ?

File: aws/cloudformation/components/database.yml.erb [hunk 8]
```diff
@@ -264,8 +264,8 @@
         # and to permit long running queries.
         InitQuery: 'SET SESSION aurora_read_replica_read_committed=ON; SET SESSION tx_isolation="READ-COMMITTED"; SET SESSION max_execution_time = 0;'
         # Ensure that connections to the reader database instances that are carrying out "reporting" queries do not
-        # exceed 5% of the total database connection capacity of a reader instance.
-        MaxConnectionsPercent: 5
+        # exceed 10% of the total database connection capacity of a reader instance.
+        MaxConnectionsPercent: 10
```
Comment (Hamms) [inline aws/cloudformation/components/database.yml.erb:268]: Is there a specific reason for this increase?

**Review by Hamms — COMMENTED** [review]
> $ bundle exec rake stack:data:validate RAILS_ENV=production ADMIN=true REDSHIFT_ADMIN_USERNAME=[REDACTED] REDSHIFT_ADMIN_PASSWORD=[REDACTED]

I'm a bit worried about us putting admin passwords in plaintext on the command line, where they will be preserved in our bash history. Is there any functionality here that can prompt the developer to input the password interactively, as with `mysql -p`?

**Comment by sureshc** [comment]
Agreed. I ran these locally in a way that did not persist to my command history, but looks like our `AWS::CloudFormation` class will interactively prompt for missing Parameters. I can use that to execute this and document it in this Pull Request as the safest way to run this stack update

https://github.com/code-dot-org/code-dot-org/blob/05620d8d651a8ab761ab47df6f551180a71a6633/lib/cdo/aws/cloud_formation.rb#L226

**Comment by sureshc** [comment]
Data Stack updated:

```
 Stack update requested, waiting for provisioning to complete...
.2024-06-28 21:56:48 UTC- MySQLDatabaseSourceEndpoint [CREATE_IN_PROGRESS]
2024-06-28 21:56:49 UTC- Tableau [UPDATE_IN_PROGRESS]
2024-06-28 21:56:50 UTC- MySQLDatabaseSourceEndpoint [CREATE_IN_PROGRESS]: Resource creation Initiated
2024-06-28 21:56:50 UTC- MySQLDatabaseSourceEndpoint [CREATE_COMPLETE]
.............................2024-06-28 22:01:28 UTC- Tableau [UPDATE_COMPLETE]
2024-06-28 22:01:31 UTC- RedshiftEndpoint [UPDATE_IN_PROGRESS]
.2024-06-28 22:01:33 UTC- RedshiftEndpoint [UPDATE_COMPLETE]
2024-06-28 22:01:34 UTC- DMSCronUserLevels [UPDATE_IN_PROGRESS]: Requested update requires the creation of a new physical resource; hence creating one.
2024-06-28 22:01:34 UTC- DMSCronUserHierarchyPii [UPDATE_IN_PROGRESS]: Requested update requires the creation of a new physical resource; hence creating one.
2024-06-28 22:01:34 UTC- DMSCronLevelSourcesPii [UPDATE_IN_PROGRESS]: Requested update requires the creation of a new physical resource; hence creating one.
2024-06-28 22:01:34 UTC- DMSCronPii [UPDATE_IN_PROGRESS]: Requested update requires the creation of a new physical resource; hence creating one.
2024-06-28 22:01:34 UTC- DMSCronUserHierarchy [UPDATE_IN_PROGRESS]: Requested update requires the creation of a new physical resource; hence creating one.
2024-06-28 22:01:35 UTC- DMSCronUserHierarchyPii [UPDATE_IN_PROGRESS]: Resource creation Initiated
2024-06-28 22:01:35 UTC- DMSCron [UPDATE_IN_PROGRESS]: Requested update requires the creation of a new physical resource; hence creating one.
2024-06-28 22:01:35 UTC- DMSCronUserLevels [UPDATE_IN_PROGRESS]: Resource creation Initiated
2024-06-28 22:01:35 UTC- DMSCronPii [UPDATE_IN_PROGRESS]: Resource creation Initiated
2024-06-28 22:01:36 UTC- DMSCronLevelSourcesPii [UPDATE_IN_PROGRESS]: Resource creation Initiated
2024-06-28 22:01:36 UTC- DMSCronUserHierarchy [UPDATE_IN_PROGRESS]: Resource creation Initiated
2024-06-28 22:01:37 UTC- DMSCron [UPDATE_IN_PROGRESS]: Resource creation Initiated
....2024-06-28 22:02:19 UTC- DMSCronUserHierarchyPii [UPDATE_COMPLETE]
2024-06-28 22:02:20 UTC- DMSCronPii [UPDATE_COMPLETE]
2024-06-28 22:02:20 UTC- DMSCronUserLevels [UPDATE_COMPLETE]
2024-06-28 22:02:20 UTC- DMSCronUserHierarchy [UPDATE_COMPLETE]
.....2024-06-28 22:03:03 UTC- DMSCronLevelSourcesPii [UPDATE_COMPLETE]
2024-06-28 22:03:06 UTC- DMSCron [UPDATE_COMPLETE]
2024-06-28 22:03:07 UTC- DATA-production [UPDATE_COMPLETE_CLEANUP_IN_PROGRESS]
2024-06-28 22:03:03 UTC- DMSCronLevelSourcesPii [UPDATE_COMPLETE]
2024-06-28 22:03:06 UTC- DMSCron [UPDATE_COMPLETE]
2024-06-28 22:03:07 UTC- DATA-production [UPDATE_COMPLETE_CLEANUP_IN_PROGRESS]

Stack update complete.
Outputs:
Finished stack:data:start (11 minutes)
```

**Pull request merged**